kubernetes/api/v1alpha1/servicelevelobjective_types.go

```diff
@@ -159,7 +164,84 @@ type Query struct {
 // ServiceLevelObjectiveStatus defines the observed state of ServiceLevelObjective.
 type ServiceLevelObjectiveStatus struct{}
 
-func (in ServiceLevelObjective) Internal() (slo.Objective, error) {
+func (in *ServiceLevelObjective) ValidateCreate() (admission.Warnings, error) {
+	return in.validate()
+}
+
+func (in *ServiceLevelObjective) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	return in.validate()
+}
+
+func (in *ServiceLevelObjective) ValidateDelete() (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (in *ServiceLevelObjective) validate() (admission.Warnings, error) {
+	var warnings []string
+
+	if in.GetName() == "" {
+		return warnings, fmt.Errorf("name must be set")
+	}
+	if in.GetNamespace() == "" {
+		warnings = append(warnings, "namespace must be set")
+	}
+
+	if in.Spec.Target == "" {
+		return warnings, fmt.Errorf("target must be set")
+	}
+	target, err := strconv.ParseFloat(in.Spec.Target, 64)
+	if err != nil {
+		return warnings, err
+	}
+	if target < 0 || target > 100 {
+		return warnings, fmt.Errorf("target must be between 0 and 100")
+	}
+	if target > 0 && target < 1 {
+		warnings = append(warnings, fmt.Sprintf("target is from 0-100 (%v), not 0-1 (%v)", 100*target, target))
+	}
+
+	if in.Spec.Window == "" {
+		return warnings, fmt.Errorf("window must be set")
+	}
+	_, err = model.ParseDuration(in.Spec.Window)
+	if err != nil {
+		return warnings, err
+	}
+
+	if in.Spec.ServiceLevelIndicator.Ratio == nil &&
+		in.Spec.ServiceLevelIndicator.Latency == nil &&
+		in.Spec.ServiceLevelIndicator.LatencyNative == nil &&
+		in.Spec.ServiceLevelIndicator.BoolGauge == nil {
+		return warnings, fmt.Errorf("one of ratio, latency, latencyNative or bool_gauge must be set")
+	}
+
+	if in.Spec.ServiceLevelIndicator.Ratio != nil {
+		ratio := in.Spec.ServiceLevelIndicator.Ratio
+		if ratio.Total.Metric == "" {
+			return warnings, fmt.Errorf("ratio total metric must be set")
+		}
+		if ratio.Errors.Metric == "" {
+			return warnings, fmt.Errorf("ratio errors metric must be set")
+		}
+
+		if ratio.Errors.Metric == ratio.Total.Metric {
```

Comment:
Is there a reason why identical queries are not checked for the other SLO types? (unless I missed them?)

Reply:
The other indicator types are not yet checked at all. First, I thought we can get the ratio in with all the surrounding boilerplate and then take care of the other indicators.

Reply:
Make sense :+1: